### PR TITLE
Allow edit access to prior reporting period

### PR DESCRIPTION
**Pull request by batemapf**
## Description

Gives the user access to a limited number of previously submitted timecards for editing (requested in #572). Typically, only one previously submitted timecard will be available; however, when a new reporting period has been created but has not yet begun, the user has access to the "current" reporting period, ~as well as the prior week~ as well as any timecards that have not been submitted and timecards for reporting periods with end dates that are seven or less days in the past. The objective is to ensure the user always has access to the current week's timecard for editing, as well as the prior week's. 

This new view is accessible via the Reports page.

In addition, this change addresses #401 by limiting access to the timecard form based on the reporting period. A user who tries to access the timecard form for a reporting period ~that is earlier than the previous reporting period~ that has an end date which is more than seven days in the past  will be redirected to their submitted data. The exception is if the user never submitted a timecard for that period. In that case, the user will be directed to their form for submission. This should only happen in instances where the user is at least two weeks late in submitting their time.

Further tweaks were made to the user's initial view, including ~adding in links to previous timecards (view-only) and~ displaying a notification that the user is up to date if all available timecard forms have been filed.

## Additional information

*Updated* Screen shot of new view:

----
screen shot 2016-12-06 at 4 39 17 pm

----

Reports page:

screen shot 2016-12-06 at 4 42 01 pm

### CODE REVIEW

**Comment by batemapf** [comment]
Basic functionality is complete. Needs tests, user input, and likely a little refactoring.

**Comment by batemapf** [comment]
Related to discussion #566 cc: @line47 @sfvictor

File: tock/hours/views.py [hunk 1]
```diff
@@ -230,11 +230,22 @@ class ReportingPeriodListView(PermissionMixin, ListView):
...
+        ).distinct().order_by('-start_date')
+        try:
+            if datetime.date.today() < completed_reporting_periods[0].start_date:
+                increment = 2
+            else:
+                increment = 1
+            context['amendable_completed_reporting_periods'] = \
+                completed_reporting_periods[:increment]
+            context['completed_reporting_periods'] = \
+                completed_reporting_periods[increment:5]
```
Comment (batemapf) [inline tock/hours/views.py]: TODO: Test that amendable reporting periods are pulled.

File: tock/hours/views.py [hunk 2]
```diff
@@ -230,11 +230,22 @@ class ReportingPeriodListView(PermissionMixin, ListView):
...
+        except IndexError:
+            context['amendable_completed_reporting_periods'] = None
```
Comment (batemapf) [inline tock/hours/views.py]: TODO: Check that a user with no reporting periods has errors handled.

File: tock/hours/views.py [hunk 3]
```diff
@@ -230,11 +230,22 @@ class ReportingPeriodListView(PermissionMixin, ListView):
...
+            context['completed_reporting_periods'] = None
```
Comment (batemapf) [inline tock/hours/views.py]: TODO: Test that correct reporting periods are returned in context.

File: tock/hours/views.py [hunk 4]
```diff
@@ -327,6 +338,43 @@ def get_object(self, queryset=None):
...
+        else:
+            increment = 1
+        last_rps = max_rps[:increment]
+
+        # If the reporting period is in the reporting period list or if the
+        # timecard has not been submitted, return the timecard form. If not,
+        # return the non-editable report view of the timecard.
+        if selected_rp in last_rps or \
+        self.object.__dict__['submitted'] == False:
+            context = self.get_context_data(object=self.object)
```
Comment (batemapf) [inline tock/hours/views.py]: TODO: Check that an unsubmitted form is _always_ available for editing.

File: tock/hours/views.py [hunk 5]
```diff
@@ -327,6 +338,43 @@ def get_object(self, queryset=None):
...
+            return self.render_to_response(context)
+        else:
+            return redirect(
+                reverse(
```
Comment (batemapf) [inline tock/hours/views.py]: TODO: Check that redirect occurs and that destination is correct.

File: tock/tock/templates/hours/reporting_period_list.html [hunk 6]
```diff
@@ -9,10 +9,41 @@
...
+					to
+					{{ reporting_period.end_date | date:"F j, Y" }}</a></li>
+			{% endfor %}
+			</ul>
+		{% endif %}
+		{% if completed_reporting_periods %}
+	<h3>Prior Reporting Periods That May Not be Amended by You</h3>
+		<ul>
+		{% for reporting_period in completed_reporting_periods %}
+			<li><a href="{% url 'reports:ReportingPeriodUserDetailView' reporting_period=reporting_period.start_date username=user.username %}">
```
Comment (batemapf) [inline tock/tock/templates/hours/reporting_period_list.html]: TODO: Test that correct URLs for reporting periods are returned where the reporting periods cannot be edited.

File: tock/tock/templates/hours/reporting_period_list.html [hunk 7]
```diff
@@ -9,10 +9,41 @@
...
+					{{ reporting_period.end_date | date:"F j, Y" }}</a></li>
+			{% endfor %}
+		{% else %}
+			<i>You are up to date!</i>
+		{% endif %}
 	</ul>
+		{% if amendable_completed_reporting_periods %}
+		<h3>Prior Reporting Periods That May be Amended by You</h3>
+			<ul>
+			{% for reporting_period in amendable_completed_reporting_periods %}
```
Comment (batemapf) [inline tock/tock/templates/hours/reporting_period_list.html]: TODO: Test that all reporting periods that are able to be amended are displayed.

File: tock/tock/templates/hours/reporting_period_list.html (hunk 6 above)
Comment (batemapf) [inline tock/tock/templates/hours/reporting_period_list.html]: Done.

File: tock/hours/views.py (hunk 5 above)
Comment (batemapf) [inline tock/hours/views.py]: Done.

File: tock/hours/views.py [hunk 8]
```diff
@@ -327,6 +338,43 @@ def get_object(self, queryset=None):
...
+
+        # Get most recent reporting period objects.
+        max_rps = ReportingPeriod.objects.all().order_by('-start_date')[:3]
+
+        # Determine increment and number of reporting periods in list.
+        if datetime.date.today() < max_rps[0].start_date:
+            increment = 2
+        else:
+            increment = 1
+        last_rps = max_rps[:increment]
```
Comment (batemapf) [inline tock/hours/views.py]: Done.

File: tock/hours/views.py (hunk 3 above)
Comment (batemapf) [inline tock/hours/views.py]: Done.

File: tock/hours/views.py (hunk 1 above)
Comment (batemapf) [inline tock/hours/views.py]: Done.

File: tock/hours/views.py [hunk 9]
```diff
@@ -230,11 +230,22 @@ class ReportingPeriodListView(PermissionMixin, ListView):
...
             ReportingPeriodListView, self).get_context_data(**kwargs)
-        context['completed_reporting_periods'] = self.queryset.filter(
+        completed_reporting_periods = self.queryset.filter(
             timecard__submitted=True,
             timecard__user=self.request.user.id
-        ).distinct().order_by('-start_date')[:5]
-
+        ).distinct().order_by('-start_date')
+        try:
+            if datetime.date.today() < completed_reporting_periods[0].start_date:
```
Comment (batemapf) [inline tock/hours/views.py]: Done.

File: tock/tock/templates/hours/reporting_period_list.html (hunk 7 above)
Comment (batemapf) [inline tock/tock/templates/hours/reporting_period_list.html]: Done via context testing and assertTemplateUsed.

**Comment by batemapf** [comment]
@annalee this one is ready for you!

**Comment by batemapf** [comment]
@maya you also should take a look since this represents a sizeable change in the initial user view!

**Comment by maya** [comment]
@batemapf can we use 2 tab spaces for the HTML template page to make reviewing easier?

Can we use sentence case for the headings per the 18F content guide: https://pages.18f.gov/content-guide/capitalization/#headings ?

The two new sections are a bit overwhelming and distract from the primary use case of filling out the current timecard for the week. We'll need to de-emphasize this.

**Comment by maya** [comment]
In "Prior Reporting Periods That May Not be Amended by You," the "See more" link just brings me to Reports page, not my personal page of past timecards. I expect to go to see all of my past timecards with this link.

**Comment by maya** [comment]
This is my first stab at trying to address these issues. What do you think?

screen shot 2016-12-05 at 11 27 05 am

**Comment by batemapf** [comment]
@maya thank you for the comments! here is where we're at:

>> In "Prior Reporting Periods That May Not be Amended by You," the "See more" link just brings me to Reports page, not my personal page of past timecards. I expect to go to see all of my past timecards with this link.

addressed this in https://github.com/18F/tock/pull/580/commits/fac563ddf5d83b772f02e550cab3416f359a0a22. 

---

>> The two new sections are a bit overwhelming and distract from the primary use case of filling out the current timecard for the week. We'll need to de-emphasize this.

i agree. we could address this a little differently and move the this functionality off of the "main" page and into something like:

screen shot 2016-12-05 at 5 20 04 pm

(which is the view created via https://github.com/18F/tock/pull/580/commits/fac563ddf5d83b772f02e550cab3416f359a0a22), but with options for each timecard to view / edit.

---

>> When I click on "Prior Reporting Periods That May be Amended by You," it brings me to the user summary for that week, but there's no way to edit it. I expect to go to a page that's like the Tock editable timesheet or this page + a "edit" link to that page.

that is pretty worrisome! you should be taken to the timecard form for that reporting period, not the reports page. will see if i can replicate it locally.

**Comment by batemapf** [comment]
@annalee unassigning you for now since we still have further to go on this one.

**Comment by maya** [comment]
>> The two new sections are a bit overwhelming and distract from the primary use case of filling out the current timecard for the week. We'll need to de-emphasize this.

>i agree. we could address this a little differently and move the this functionality off of the "main" page and into something like:

Yeah maybe we remove that section and replace it with a link, "View all timecards." Also, I think we may want to run those headings through the writing lab.

**Comment by batemapf** [comment]
@maya did a pretty big overhaul of this which is ready for your review. see updated comment! the logic is reworked to allow edit access to:
- any unsubmitted timecards;
- any submitted timecard for reporting periods that have not ended; and
- any submitted timecards for reporting periods that have ended, but seven or less days ago.

**Comment by maya** [comment]
@batemapf re: UX, this works for me! 

Is the goal to keep it a little hidden so people aren't encouraged to make changes to last week's timecard often? This relies on you knowing where to go to make the change: "Reports" > "List of all timecards" which can be learned, we'll have to see how people use it. Something more straight forward like: `Edit last week's time card` on the homepage is more obvious, but may encourage people to use the feature more than we want.

**Comment by batemapf** [comment]
@maya thanks! unassigning...

re: goal of moving it off the main page is not to hide the functionality, but to keep the main page as sparse as possible. of course, if we get feedback that users are looking for more immediate access to past time cards, we can easily drop a link to the view on the main page.

@annalee you are up!

**Pull request closed without merging**